src/main/haskell/kore/src/Kore/Step/MultiAnd.hs

```diff
@@ -0,0 +1,136 @@
+{-|
+Module      : Kore.Step.MultiAnd
+Description : Data structures and functions for manipulating
+              And with any number of children.
+Copyright   : (c) Runtime Verification, 2019
+License     : NCSA
+Maintainer  : virgil.serbanuta@runtimeverification.com
+Stability   : experimental
+Portability : portable
+-}
+module Kore.Step.MultiAnd
+    ( MultiAnd
+    , extractPatterns
+    , make
+    ) where
+
+import           Control.DeepSeq
+                 ( NFData )
+import qualified Data.Set as Set
+import           GHC.Generics
+                 ( Generic )
+
+import Kore.TopBottom
+       ( TopBottom (..) )
+
+{-| 'MultiAnd' is a Matching logic and of its children
+
+-}
+{- TODO (virgil): Make 'getMultiAnd' a non-empty list ("Data.NonEmpty").
+
+An empty 'MultiAnd' corresponding to 'Top' actually discards information
+about the sort of its child patterns! That is a problem for simplification,
+which should preserve pattern sorts.
+
+A non-empty 'MultiAnd' would also have a nice symmetry between 'Top' and
+'Bottom' patterns.
+-}
+newtype MultiAnd child = MultiAnd { getMultiAnd :: [child] }
+  deriving
+    (Applicative, Eq, Foldable, Functor, Generic, Monad, Ord, Show, Traversable)
+
+instance NFData child => NFData (MultiAnd child)
+
+instance TopBottom child => TopBottom (MultiAnd child)
+  where
+    isTop (MultiAnd []) = True
+    isTop _ = False
+    isBottom (MultiAnd [child]) = isBottom child
+    isBottom _ = False
+
+{-| 'AndBool' is an some sort of Bool data type used when evaluating things
+inside an 'MultiAnd'.
+-}
+-- TODO(virgil): Refactor, this is the same as OrBool
```

Comment:
You don't need to address it in this pull request, but why not `Maybe Bool` for both?

Reply:
I usually prefer using specific types whenever that does not make other things too hard. In particular, this type is actually  supposed to be `Top, Bottom, OtherPattern`, which, IMO, says much more than a `Maybe Bool`, for historical reasons, the names are a bit different. I added a TODO.